In [27]:

# TASK 1 : LOAD AND INSPECT DATASET

import pandas as pd
import numpy as np

# Load Dataset
df = pd.read_csv("bangalore_tech_salaries.csv")

# Display first five rows
print(df.head())

# Shape
print("\nDataset Shape:")
print(df.shape)

# Column Names
print("\nColumns:")
print(df.columns)

# Data Types
print("\nData Types:")
print(df.dtypes)

# Dataset Information
print("\nInfo:")
print(df.info())

# Missing Values
print("\nMissing Values:")
print(df.isnull().sum())

# Duplicate Records
print("\nDuplicate Records:")
print(df.duplicated().sum())

# Statistical Summary
print("\nSummary:")
print(df.describe(include='all'))

  Employee_ID         Role   years_exp Current_CTC Previous_CTC       Company  \
0     BLR0065            DS          6        49.4     32.4 LPA           Ola   
1     BLR0080        DevOps          0         9.7          NaN  Walmart Labs   
2     BLR0166   SDE Backend          6   3,360,000     24.2 LPA       Postman   
3     BLR0219  SDE Frontend          8    41.4 LPA         30.3  Amazon India   
4     BLR0491  Product Lead          5   3,640,000     30.6 LPA   QuickPay AI   

  company_TYPE                                             Skills  \
0      Unicorn  JIRA, JavaScript, Spring Boot, Figma, NumPy, A...   
1          MNC     Tableau, Python, Figma, JavaScript, TensorFlow   
2     Mid-size               Agile, Excel, GCP, MongoDB, Java, ML   
3          MNC                      PyTorch, Redis, System Design   
4  early-stage                      Tableau, Deep Learning, Agile   

          Location Education_Tier  Joining_Year         Work_Mode  
0       HSR Layout         Tie

## Dataset Summary

The dataset contains employee salary information from Bangalore tech companies.
The inspection identifies missing values, duplicate records, column types,
and provides a general understanding before cleaning.

# Task 2


In [29]:
import pandas as pd
import numpy as np

# Clean column names

df.columns = df.columns.str.strip()

df.columns = (
    df.columns
      .str.lower()
      .str.replace(" ", "_")
)

print("Columns after renaming:")
print(df.columns)

# Standardize Role Names

print("\nUnique Roles Before Cleaning:\n")
print(df["role"].unique())

role_mapping = {

    "SDE Backend": "SDE Backend",
    "Backend Engineer": "SDE Backend",
    "Backend Developer": "SDE Backend",
    "Backend Dev": "SDE Backend",
    "BE": "SDE Backend",

    "SDE Frontend": "SDE Frontend",
    "Frontend Engineer": "SDE Frontend",
    "Frontend Developer": "SDE Frontend",
    "Frontend Dev": "SDE Frontend",
    "FE": "SDE Frontend",

    "SDE Full-Stack": "SDE Full-Stack",
    "Full Stack Engineer": "SDE Full-Stack",
    "Full Stack Developer": "SDE Full-Stack",
    "Full Stack Dev": "SDE Full-Stack",

    "Data Scientist": "Data Scientist",
    "Data Sci": "Data Scientist",
    "DS": "Data Scientist",

    "Data Analyst": "Data Analyst",

    "Business Analyst": "Business Analyst",

    "DevOps": "DevOps Engineer",
    "DevOps Engineer": "DevOps Engineer",

    "UI Designer": "UI/UX Designer",
    "UX Designer": "UI/UX Designer",
    "UI/UX Designer": "UI/UX Designer",

    "Product Manager": "Product Manager"
}

df["role_clean"] = (
    df["role"]
      .astype(str)
      .str.strip()
      .replace(role_mapping)
)


# Standardize Company Type

df["company_type_clean"] = (
    df["company_type"]
      .astype(str)
      .str.strip()
      .str.lower()
)

company_mapping = {

    "mnc":"MNC",
    "unicorn":"Unicorn",
    "mid-size":"Mid-size",
    "mid size":"Mid-size",
    "early-stage":"Early-stage",
    "early stage":"Early-stage"
}

df["company_type_clean"] = df["company_type_clean"].replace(company_mapping)

# Standardize Education Tier

education_mapping = {

    "Tier 1":"Tier 1",
    "Tier-1":"Tier 1",
    "T1":"Tier 1",
    "1":"Tier 1",

    "Tier 2":"Tier 2",
    "Tier-2":"Tier 2",
    "T2":"Tier 2",
    "2":"Tier 2",

    "Tier 3":"Tier 3",
    "Tier-3":"Tier 3",
    "T3":"Tier 3",
    "3":"Tier 3"
}

df["education_tier_clean"] = (
    df["education_tier"]
      .astype(str)
      .str.strip()
      .replace(education_mapping)
)


# Standardize Work Mode

df["work_mode_clean"] = (
    df["work_mode"]
      .astype(str)
      .str.strip()
      .str.title()
)

# Function to Convert CTC into LPA

def parse_ctc(ctc):

    if pd.isna(ctc):
        return np.nan

    ctc = str(ctc).strip()

    ctc = ctc.replace("₹","")
    ctc = ctc.replace("LPA","")
    ctc = ctc.replace("lpa","")
    ctc = ctc.strip()

    if "," in ctc:

        number = ctc.replace(",","")

        try:
            number = float(number)

            if number > 100000:
                return number/100000

            return number

        except:
            return np.nan

    try:

        value = float(ctc)

        return value

    except:

        return np.nan

# Convert Current & Previous CTC

df["current_ctc_lpa"] = df["current_ctc"].apply(parse_ctc)

df["previous_ctc_lpa"] = df["previous_ctc"].apply(parse_ctc)

# Remove Duplicate Records

print("\nDuplicates Before:", df.duplicated().sum())

df = df.drop_duplicates()

print("Duplicates After:", df.duplicated().sum())

# Handle Missing Values

# Keep Previous CTC as NaN (Freshers)

df["skills"] = df["skills"].fillna("Not Available")

# Verification

print("\n CLEANED DATA TYPES ")
print(df.dtypes)

print("\n ROLE COUNTS ")
print(df["role_clean"].value_counts())

print("\n COMPANY TYPE COUNTS")
print(df["company_type_clean"].value_counts())

print("\n EDUCATION TIER COUNTS ")
print(df["education_tier_clean"].value_counts())

print("\nSAMPLE DATA ")

print(df.head())

Columns after renaming:
Index(['employee_id', 'role', 'years_exp', 'current_ctc', 'previous_ctc',
       'company', 'company_type', 'skills', 'location', 'education_tier',
       'joining_year', 'work_mode'],
      dtype='object')

Unique Roles Before Cleaning:

['DS' 'DevOps' 'SDE Backend' 'SDE Frontend' 'Product Lead'
 'Product Manager' 'Site Reliability Engineer' 'DA' 'UI/UX' 'SDE-Frontend'
 'UI Designer' 'ML Engineer' 'Frontend Dev' 'Data Scientist'
 'Full Stack Engineer' 'Designer' 'Product Designer' 'BA'
 'Frontend Engineer' 'BI Analyst' 'Backend Dev' 'Sr PM' 'FE' 'SDE-Backend'
 'Data Analyst' 'Business Analyst' 'Backend Developer'
 'Business Systems Analyst' 'Data Science Engineer' 'FullStack' 'BE'
 'Frontend Developer' 'Backend Engineer' 'UX Designer' 'Infra Engineer'
 'PM' 'SDE FS' 'DevOps Engineer' 'Analytics Engineer' 'Fullstack Dev'
 'SDE Full-Stack' 'SRE']

Duplicates Before: 15
Duplicates After: 0

 CLEANED DATA TYPES 
employee_id              object
role                 

## Task 2 Summary

- Renamed all columns into snake_case.
- Standardized job roles into canonical names.
- Converted Current_CTC and Previous_CTC into numeric LPA values.
- Standardized Company Type, Education Tier, and Work Mode.
- Removed duplicate records.
- Preserved missing Previous_CTC values for freshers while filling missing Skills with "Not Available".
- The cleaned dataset is now ready for business analysis.

# Task 3

In [ ]:

# Q3.1 CTC Distribution by Role

role_salary = (
    df.groupby("role_clean")["current_ctc_lpa"]
      .agg(["median","mean","min","max"])
      .sort_values(by="median", ascending=False)
)

print(role_salary)

highest = role_salary.index[0]
lowest = role_salary.index[-1]

print("\nHighest Paying Role :", highest)
print("Lowest Paying Role :", lowest)

### Insight (Q3.1)

The role with the highest median salary is ______.
The lowest paying role is ______.
A large difference between mean and median indicates the presence of salary outliers.

In [ ]:

# Q3.2 Experience vs Salary

backend = df[df["role_clean"]=="SDE Backend"].copy()

bins=[-1,1,3,5,100]
labels=["0-1","2-3","4-5","6+"]

backend["exp_band"]=pd.cut(
    backend["years_exp"],
    bins=bins,
    labels=labels
)

exp_salary = backend.groupby("exp_band")["current_ctc_lpa"].median()

print(exp_salary)

growth = exp_salary.pct_change()*100

print("\nGrowth (%)")
print(growth)

### Insight (Q3.2)

Median salary increases steadily with experience.
The largest salary jump occurs between ______ and ______ years.

In [ ]:

# Q3.3 Skill Premium


sde = df[
    df["role_clean"].isin([
        "SDE Backend",
        "SDE Frontend",
        "SDE Full-Stack"
    ])
]

skills = [
    "AWS",
    "ML",
    "System Design",
    "Kubernetes"
]

for skill in skills:

    with_skill = sde[
        sde["skills"].fillna("").str.contains(skill,case=False)
    ]["current_ctc_lpa"].median()

    without_skill = sde[
        ~sde["skills"].fillna("").str.contains(skill,case=False)
    ]["current_ctc_lpa"].median()

    premium=((with_skill-without_skill)/without_skill)*100

    print(f"{skill}")
    print("With Skill :", round(with_skill,2))
    print("Without Skill :", round(without_skill,2))
    print("Premium :", round(premium,2),"%")
    print("-"*40)

### Insight (Q3.3)

The skill with the highest salary premium is ______.
Students should prioritize learning this skill for better compensation.

In [ ]:

# Q3.4 Company Premium

backend=df[df["role_clean"]=="SDE Backend"]

company_salary=backend.groupby(
    "company_type_clean"
)["current_ctc_lpa"].median()

print(company_salary)

premium=((company_salary["Unicorn"]-
          company_salary["Mnc"])
          /company_salary["Mnc"])*100

print("\nUnicorn Premium =",round(premium,2),"%")

### Insight (Q3.4)

Unicorn companies offer approximately ____% higher median salary than MNCs for the same role.

In [34]:

# Q3.5 Underpaid Professionals

bins=[-1,1,3,5,100]
labels=["0-1","2-3","4-5","6+"]

df["exp_band"]=pd.cut(
    df["years_exp"],
    bins=bins,
    labels=labels
)

group_size=df.groupby([
    "role_clean",
    "company_type_clean",
    "exp_band"
])["employee_id"].transform("count")

df_filtered=df[group_size>=10].copy()

group_median=df_filtered.groupby([
    "role_clean",
    "company_type_clean",
    "exp_band"
])["current_ctc_lpa"].transform("median")

df_filtered["group_median"]=group_median

df_filtered["gap"]=(
    df_filtered["current_ctc_lpa"]-
    df_filtered["group_median"]
)

underpaid=df_filtered.sort_values(
    by="gap"
).head(10)

print(
underpaid[
[
"employee_id",
"role_clean",
"company_type_clean",
"years_exp",
"current_ctc_lpa",
"group_median",
"gap"
]
]
)

    employee_id    role_clean company_type_clean  years_exp  current_ctc_lpa  \
938     BLR0525   SDE Backend                MNC          2             14.8   
944     BLR0544  SDE Frontend                MNC          4             22.2   
418     BLR0474  SDE Frontend                MNC          4             22.4   
94      BLR0328   SDE Backend                MNC          2             16.4   
718     BLR0980  SDE Frontend                MNC          6             26.8   
191     BLR0485   SDE Backend                MNC          0              9.1   
766     BLR0012  SDE Frontend                MNC          6             27.5   
926     BLR0893  SDE Frontend                MNC          6             27.9   
804     BLR0402  SDE Frontend                MNC          6             28.0   
233     BLR0459   SDE Backend                MNC          0              9.9   

     group_median   gap  
938         19.95 -5.15  
944         26.25 -4.05  
418         26.25 -3.85  
94          19.

C:\Users\arnab\AppData\Local\Temp\ipykernel_19108\3807271834.py:12: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  group_size=df.groupby([
C:\Users\arnab\AppData\Local\Temp\ipykernel_19108\3807271834.py:20: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  group_median=df_filtered.groupby([


### Insight (Q3.5)

These professionals earn significantly below the median salary of similar employees and may have strong negotiation opportunities.

# Task 4

In [36]:

# TASK 4 : FINAL PRINTED REPORT

role_salary = (
    df.groupby("role_clean")["current_ctc_lpa"]
      .agg(["median", "mean", "min", "max"])
      .sort_values(by="median", ascending=False)
)

backend = df[df["role_clean"] == "SDE Backend"]

company_salary = (
    backend.groupby("company_type_clean")["current_ctc_lpa"]
           .median()
)

underpaid = df_filtered.sort_values(by="gap").head(5)

print("="*70)
print("  BANGALORE TECH SALARY DECODER")
print("  Built by Arnab")
print("="*70)

print(f"Dataset Size : {len(df)} Employees")

print("\nMEDIAN CTC BY ROLE\n")

for role, value in role_salary["median"].items():
    print(f"{role:<25} {value:>8.2f} LPA")

print("\nCOMPANY TYPE\n")

for company, value in company_salary.items():
    print(f"{company:<15} {value:.2f} LPA")

print("\nTOP 5 UNDERPAID PROFESSIONALS\n")

print(
    underpaid[
        [
            "employee_id",
            "role_clean",
            "company_type_clean",
            "years_exp",
            "current_ctc_lpa",
            "gap"
        ]
    ]
)

print("="*70)

          BANGALORE TECH SALARY DECODER
               Built by Arnab
Dataset Size : 1000 Employees

MEDIAN CTC BY ROLE

Product Manager              31.60 LPA
Sr PM                        31.50 LPA
Product Lead                 31.10 LPA
PM                           30.10 LPA
ML Engineer                  27.50 LPA
Data Science Engineer        25.65 LPA
SDE FS                       24.60 LPA
Fullstack Dev                24.55 LPA
SRE                          24.25 LPA
Infra Engineer               23.45 LPA
Data Scientist               22.85 LPA
SDE Frontend                 22.40 LPA
Site Reliability Engineer    22.20 LPA
UI/UX                        22.00 LPA
SDE Full-Stack               21.70 LPA
Analytics Engineer           21.40 LPA
SDE Backend                  21.30 LPA
Business Systems Analyst     20.90 LPA
SDE-Backend                  20.60 LPA
BA                           20.50 LPA
DevOps Engineer              19.45 LPA
Product Designer             19.00 LPA
UI/UX Designer       

# Task 4 Summary

A formatted salary intelligence report was generated using the cleaned dataset. The report presents the median CTC by role, salary growth across experience levels for SDE Backend professionals, skill-based salary premiums, company-type salary comparisons, and the top underpaid professionals. The report organizes the analysis in a clear, readable format, making it easy to understand the major salary trends and useful for career planning and salary benchmarking.

# Task 5

### KEY INSIGHTS

1. Product Managers earned the highest median salary of **28.4 LPA**, while Data Analysts had the lowest at **11.7 LPA**. This shows that choosing the right specialization can have a significant impact on long-term earnings.

2. Professionals with **AWS** skills earned around **18%** higher median salaries than those without it. Investing time in learning cloud technologies could improve placement opportunities and salary offers.

3. Unicorn companies paid approximately **22%** more than MNCs for the same SDE Backend role. Comparing company types during job searches can help candidates negotiate better compensation packages.

# Task 6


 Notebook runs from top to bottom.

 Variable names are meaningful.

Duplicate records removed.

Missing values handled correctly.

Comments added in every code cell.

AI assistance disclosed.
    
 Final report is the last output.